src/box/tuple_constraint_fkey.c

```diff
@@ -114,12 +127,127 @@ field_foreign_key_failed(const struct tuple_constraint *constr,
 			 const struct tuple_field *field,
 			 const char *message)
 {
-	const char *field_path = tuple_field_path(field, constr->space->format);
-	struct error *err = diag_set(ClientError, ER_FIELD_FOREIGN_KEY_FAILED,
-				     constr->def.name, field_path, message);
+	struct error *err;
+	const char *field_path = NULL;
+	if (field != NULL) {
+		field_path = tuple_field_path(field, constr->space->format);
+		err = diag_set(ClientError, ER_FIELD_FOREIGN_KEY_FAILED,
+			       constr->def.name, field_path, message);
+	} else {
+		err = diag_set(ClientError, ER_COMPLEX_FOREIGN_KEY_FAILED,
+			       constr->def.name, message);
+	}
 	error_set_str(err, "name", constr->def.name);
-	error_set_str(err, "field_path", field_path);
-	error_set_uint(err, "field_id", field->id);
+	if (field != NULL) {
+		error_set_str(err, "field_path", field_path);
+		error_set_uint(err, "field_id", field->id);
+	}
+}
+
+/**
+ * Auxiliary data structure that is used for complex key extraction from tuple.
+ */
+struct key_info {
+	/** Index of key part in key definition. */
+	uint32_t index_order;
+	/** Field number of key part. */
+	uint32_t field_no;
+	/** Msgpack data of that part in tuple. */
+	const char *mp_data;
+	/** Size of msgpack data of that part in tuple. */
+	size_t mp_data_size;
+};
+
+/** Sort by index_order compare function. */
+static int
+key_info_by_order(const void *ptr1, const void *ptr2)
+{
+	const struct key_info *info1 = (const struct key_info *)ptr1;
+	const struct key_info *info2 = (const struct key_info *)ptr2;
+	return info1->index_order < info2->index_order ? -1 :
+	       info1->index_order > info2->index_order;
+}
+
+/** Sort by field_no compare function. */
+static int
+key_info_by_field_no(const void *ptr1, const void *ptr2)
+{
+	const struct key_info *info1 = (const struct key_info *)ptr1;
+	const struct key_info *info2 = (const struct key_info *)ptr2;
+	return info1->field_no < info2->field_no ? -1 :
+	       info1->field_no > info2->field_no;
+}
+
+/**
+ * Get of extract key for foreign index from local tuple by given as @a mp_data.
+ * Simply return mp_data for field foreign key - it is the field itself.
+ * For complex foreign keys collect field in one contiguous buffer.
+ * Try to place resulting key in @a *buffer, that must be a buffer of size
+ * COMPLEX_KEY_BUFFER_SIZE. If there's not enough space - allocate needed
+ * using xmalloc - that pointer is returned via @a buffer. Thus if the pointer
+ * is changed - a user of that function must free() the buffer after usage.
+ * Return pointer to ready-to-use key in any case.
+ */
+static const char *
+get_or_extract_key_mp(const struct tuple_constraint *constr,
```

Comment:
Is it possible to use the existing `key_def` infrastructure for this (`tuple_extract_key`)?

Reply:
We decided that 'no' offline.